include/bout/index_derivs.hxx

```diff
@@ -0,0 +1,789 @@
+
+    auto& derivativeRegister = DerivativeStore<FieldType>::getInstance();
+
+    switch (Method{}.meta.derivType) {
+    case (DERIV::Standard):
+    case (DERIV::StandardSecond):
+    case (DERIV::StandardFourth): {
+      if (nGuards == 1) {
+        const auto theFunc = std::bind(
+            // Method to store in function
+            &Method::template standard<Direction::value, Stagger::value, 1, FieldType>,
+            // Arguments -- first is hidden this of type-bound, others are placeholders
+            // for input field, output field, region
+            Method{}, _1, _2, _3);
+        derivativeRegister.template registerDerivative<Direction, Stagger, Method>(
+            theFunc);
+      } else {
+        const auto theFunc = std::bind(
+            // Method to store in function
+            &Method::template standard<Direction::value, Stagger::value, 2, FieldType>,
+            // Arguments -- first is hidden this of type-bound, others are placeholders
+            // for input field, output field, region
+            Method{}, _1, _2, _3);
+        derivativeRegister.template registerDerivative<Direction, Stagger, Method>(
+            theFunc);
+      }
+      break;
+    }
+    case (DERIV::Upwind):
+    case (DERIV::Flux): {
+      if (nGuards == 1) {
+        const auto theFunc = std::bind(
+            // Method to store in function
+            &Method::template upwindOrFlux<Direction::value, Stagger::value, 1,
+                                           FieldType>,
+            // Arguments -- first is hidden this of type-bound, others are placeholders
+            // for input field, output field, region
+            Method{}, _1, _2, _3, _4);
+        derivativeRegister.template registerDerivative<Direction, Stagger, Method>(
+            theFunc);
+      } else {
+        const auto theFunc = std::bind(
+            // Method to store in function
+            &Method::template upwindOrFlux<Direction::value, Stagger::value, 2,
+                                           FieldType>,
+            // Arguments -- first is hidden this of type-bound, others are placeholders
+            // for input field, output field, region
+            Method{}, _1, _2, _3, _4);
+        derivativeRegister.template registerDerivative<Direction, Stagger, Method>(
+            theFunc);
+      }
+      break;
+    }
+    default:
+      throw BoutException("Unhandled derivative method in registerMethod.");
+    };
+  }
+};
+
+/// Some helper defines for now that allow us to wrap up enums
+/// and the specific methods.
+#define e(family, value) enumWrapper<family, family::value>
+
+/////////////////////////////////////////////////////////////////////////////////
+/// Here's an example of registering a couple of DerivativeType methods
+/// at once for no staggering
+/////////////////////////////////////////////////////////////////////////////////
+
+// Could use Ben's magic macro for thing here to register multiple routines at once
```

Comment:
Could do, but probably shouldn't. There is already a near-explosive concentration of magic in this part of the code, so probably a wise choice to keep it written out. Maybe once thaum levels have decayed a little we can use it here.